jokusoramame/plugins/gambling.py

```diff
@@ -0,0 +1,210 @@
+#   - use
+#   - inventory
+
+import random
+from collections import namedtuple
+from typing import List
+
+import numpy as np
+import tabulate
+from curious import Embed, Guild, Member
+from curious.commands import Context, Plugin, command
+from curious.commands.decorators import ratelimit
+from curious.ext.paginator import ReactionsPaginator
+
+from jokusoramame.db.tables import UserBalance
+from jokusoramame.utils import chunked
+
+BAD_RESPONSES = [
+    '\N{FIRE} Your bank account went up in flames and you lost **{0} :̶.̶|̶:̶;̶**.',
+    '\N{SHOPPING TROLLEY} You spend **{0} :̶.̶|̶:̶;̶** at the supermarket.',
+    '\N{GHOST} A ghost pops up out of nowhere and scares you. You drop **{0} :̶.̶|̶:̶;̶**.',
+    '\N{HAMMER AND PICK} _All that is solid melts into air._ (You lost **{0} :̶.̶|̶:̶;̶**.)'
+]
+
+GOOD_RESPONSES = [
+    '\N{FIRST PLACE MEDAL} You win first place in the Money Making Race™ and gain **{0} :̶.̶|̶:̶;̶**.',
+    '\N{SLOT MACHINE} You have a gambling addiction and win **{0} :̶.̶|̶:̶;̶**.',
+    '\N{GEM STONE} You find a diamond and sell it for **{0} :̶.̶|̶:̶;̶**.'
+]
+
+
+class Gambling(Plugin):
+    """
+    Plugin for gambling related commands.
+    """
+
+    async def ensure_balance(self, member: Member = None):
+        """
+        Return a balance associated to the member and corresponding guild. If one does not exist it is created.
+
+        :param member: Member to ensure balance for.
+        """
+        async with self.client.db.get_session() as sess:
+            balance = await sess.select.from_(UserBalance) \
+                .where(UserBalance.user_id.eq(member.id) & UserBalance.guild_id.eq(member.guild.id)) \
+                .first()
+
+            if balance is None:
+                balance = UserBalance(guild_id=member.guild_id, user_id=member.id, money=0)
+                await sess.add(balance)
+
+            return balance
+
+    async def construct_leaderboard(self, guild: Guild, *, mode: str) -> List[str]:
+        """
+        Query the database for the balance of each member in a guild, and sort it appropriately.
+
+        A table follows this format:
+        |   POS | User    |   Money |
+        |-------+---------+---------|
+        |     1 | name    |       0 |
+
+        each table contains 10 entries
+
+        :param guild: The guild to construct the leaderboard for.
+        :param mode: Tells if the results should be ordered in ascending or descending order.
+        :return: A list of formatted tables.
+        """
+        order_by = {
+            'top': UserBalance.money.desc(),
+            'bottom': UserBalance.money.asc()
+        }
+
+        async with self.client.db.get_session() as sess:
+            query = sess.select(UserBalance) \
+                .where(UserBalance.guild_id.eq(guild.id)) \
+                .order_by(order_by[mode])
+
+            rows = await query.all()
+            rows = await rows.flatten()
+
+        entry = namedtuple('entry', 'position name money')
```

Comment:
Should be at the class level rather than doing the entire exec() cost every call.